src/algebra/commute.lean

```diff
@@ -0,0 +1,230 @@
+/-
+Copyright (c) 2019 Neil Strickland. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Neil Strickland
+
+Commuting pairs of elements in not-necessarily-commutative monoids,
+groups and (semi) rings.  Centralizers as submonoids, subgroups or
+subrings.
+-/
+
+import algebra.group_power
+  group_theory.submonoid group_theory.subgroup
+  ring_theory.subring
+
+section monoid
+
+variables {M : Type*} [monoid M]
+
+@[reducible] def commute (a b : M) : Prop := a * b = b * a
```

Comment:
Should we generalize the first part of this section to a `semigroup`?

Reply:
This is mostly done, except that mathlib does not have `is_subsemigroup` and I have not added it.